### PR TITLE
feat(common): add functions for url encoding and decoding

### PR DESCRIPTION
**Pull request by scotthart**
This functionality exists in libcurl, but requires creating a curl handle, which we're trying to better encapsulate. I've put off rolling my own prior to this, but in order to url encode object names for storage it's necessary. This implementation matches that of curl_easy_escape which escapes all printable ASCII characters except for [a-zA-Z0-9-_.~].

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9021)

### CODE REVIEW

File: google/cloud/url_encoded_test.cc [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+using ::testing::IsEmpty;
+
+TEST(UrlEncodeString, AllPrintableAsciiCharacters) {
+  std::string escaped_printable_ascii_chars =
+      R"(%20%21%22%23%24%25%26%27%28%29%2A%2B%2C-.%2F0123456789%3A%3B%3C%3D%3E%3F%40ABCDEFGHIJKLMNOPQRSTUVWXYZ%5B%5C%5D%5E_%60abcdefghijklmnopqrstuvwxyz%7B%7C%7D~)";
+  std::string printable_ascii_chars =
+      R"( !"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~)";
+  auto result = UrlEncode(printable_ascii_chars);
+  EXPECT_EQ(result, escaped_printable_ascii_chars);
+}
```
Comment (devbww) [inline google/cloud/url_encoded_test.cc:32]: Perhaps combining each encode/decode test pair into a single "round trip" test would be a feature?

File: google/cloud/url_encoded.cc [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    "A",   "B",   "C",   "D",   "E",   "F",   "G",   "H",   "I",   "J",   "K",
+    "L",   "M",   "N",   "O",   "P",   "Q",   "R",   "S",   "T",   "U",   "V",
+    "W",   "X",   "Y",   "Z",   "%5B", "%5C", "%5D", "%5E", "_",   "%60", "a",
+    "b",   "c",   "d",   "e",   "f",   "g",   "h",   "i",   "j",   "k",   "l",
+    "m",   "n",   "o",   "p",   "q",   "r",   "s",   "t",   "u",   "v",   "w",
+    "x",   "y",   "z",   "%7B", "%7C", "%7D", "~"};
+}  // namespace
+
+std::string UrlEncode(std::string const& input) {
+  if (input.empty()) return {};
```
Comment (devbww) [inline google/cloud/url_encoded.cc:36]: Why special-case empty input?

File: google/cloud/url_encoded.cc [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  std::stringstream encoded;
+  for (auto const& c : input) {
+    encoded << kEncode[c - 0x20];
+  }
+  return encoded.str();
+}
+
+std::string UrlDecode(std::string const& input) {
+  if (input.empty()) return {};
```
Comment (devbww) [inline google/cloud/url_encoded.cc:45]: Ditto on empty input.

File: google/cloud/url_encoded.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#include "google/cloud/version.h"
+#include <string>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+// Encodes the given string and returns a new string such that all input
+// characters that are not in [a-zA-z-._~] are converted to their "URL escaped"
+// version "%NN" where NN is the ASCII hex value.
```
Comment (devbww) [inline google/cloud/url_encoded.h:27]: Do we need to say anything about what happens should an input character be outside [32 ... 126]?

File: google/cloud/url_encoded.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+std::string UrlEncode(std::string const& input);
+
+// Decodes the given string, interpreting any "%NN" sequence as an "URL escaped"
+// character and substitutes the ASCII character corresponding to hex value NN.
```
Comment (devbww) [inline google/cloud/url_encoded.h:31]: Do we need to say anything about what happens should the input be malformed (e.g., a `%` not followed by two hex digits)?

File: google/cloud/url_encoded.cc [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    "L",   "M",   "N",   "O",   "P",   "Q",   "R",   "S",   "T",   "U",   "V",
+    "W",   "X",   "Y",   "Z",   "%5B", "%5C", "%5D", "%5E", "_",   "%60", "a",
+    "b",   "c",   "d",   "e",   "f",   "g",   "h",   "i",   "j",   "k",   "l",
+    "m",   "n",   "o",   "p",   "q",   "r",   "s",   "t",   "u",   "v",   "w",
+    "x",   "y",   "z",   "%7B", "%7C", "%7D", "~"};
+}  // namespace
+
+std::string UrlEncode(std::string const& input) {
+  if (input.empty()) return {};
+  std::stringstream encoded;
```
Comment (coryan) [inline google/cloud/url_encoded.cc:37]: I would think `std::string` and `operator+=` would be enough for this?

File: google/cloud/url_encoded.cc [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  for (auto const& c : input) {
+    encoded << kEncode[c - 0x20];
```
Comment (coryan) [inline google/cloud/url_encoded.cc:39]: maybe `static_assert(' ' == 0x20, "This library is only designed to work with ASCII-based character sets")`?  The code could land in a EBCDIC system someday :fear:

File: google/cloud/url_encoded.h (hunk 4 above)
Comment (coryan) [reply google/cloud/url_encoded.h:27]: I think the answer is yes, for example, GCS object names are UTF-8, and they can appear in the URL.

**Review by coryan — CHANGES_REQUESTED** [review]

File: google/cloud/url_encoded.cc (hunk 7 above)
Comment (devbww) [reply google/cloud/url_encoded.cc:39]: You might consider something like ...

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/internal/base64_transforms.cc#L27-L37

File: google/cloud/url_encoded.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_URL_ENCODED_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_URL_ENCODED_H
+
+#include "google/cloud/version.h"
+#include <string>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
```
Comment (devbww) [inline google/cloud/url_encoded.h:23]: Q.  Does this need to be externally visible?

**Pull request closed without merging**